cores/esp8266/PolledTimeout.h

```diff
@@ -45,19 +47,78 @@ struct YieldOrSkip
   static void execute() {delay(0);}
 };
 
+template <unsigned long delayMs>
+struct YieldAndDelayMs
+{
+  static void execute() {delay(delayMs);}
+};
+
 } //YieldPolicy
 
+namespace TimePolicy
+{
+
+struct TimeSourceMillis
+{
+  // time policy in milli-seconds based on millis()
+
+  using timeType = decltype(millis());
+  static timeType time() {return millis();}
+  static constexpr timeType ticksPerSecond    = 1000;
+  static constexpr timeType ticksPerSecondMax = 1000;
+};
+
+struct TimeSourceCycles
+{
+  // time policy based on ESP.getCycleCount()
+  // this particular time measurement is intended to be called very often
+  // (every loop, every yield)
+
+  using timeType = decltype(ESP.getCycleCount());
+  static timeType time() {return ESP.getCycleCount();}
+  static constexpr timeType ticksPerSecond    = F_CPU;
+  static constexpr timeType ticksPerSecondMax = 160000000; // Mhz
+};
+
+template <typename TimeSourceType, unsigned long long second_th, unsigned long long rangeCompensate = 1>
+struct TimeUnit
+{
+  using timeType = typename TimeSourceType::timeType;
+
+  static constexpr timeType user2UnitMultiplierMax = (TimeSourceType::ticksPerSecondMax * rangeCompensate) / second_th;
+  static constexpr timeType user2UnitMultiplier    = (TimeSourceType::ticksPerSecond    * rangeCompensate) / second_th;
+  static constexpr timeType user2UnitDivider       = rangeCompensate;
+  static constexpr timeType timeMax                = (std::numeric_limits<timeType>::max() - 1) / user2UnitMultiplierMax;
+  static constexpr timeType neverExpires           = std::numeric_limits<timeType>::max();
```

Comment:
neverExpires and alwaysExpired are not specific to the TimeUnit policy. I suggest removing them from here and defining them directly in in timeoutTemplate